lib/SILOptimizer/Utils/Devirtualize.cpp

```diff
@@ -1094,7 +1094,48 @@ static bool canDevirtualizeWitnessMethod(ApplySite applySite) {
     return false;
   }
 
-  return true;
+  // FIXME: devirtualizeWitnessMethod below does not support cases with
+  // covariant 'Self' nested inside a collection type,
+  // like '[Self]' or '[* : Self]'.
+  const Type interfaceTy = wmi->getMember()
+                               .getDecl()
+                               ->getInterfaceType()
```

Comment:
This is just getResultInterfaceType()

Reply:
`getResultInterfaceType()` is not a member of `AbstractFunctionDecl`, only `FuncDecl`

Reply:
ConstructorDecl and DestructorDecl cannot return a collection of Self though, so FuncDecl is the only relevant case here.

Reply:
We need to exclude other covariant positions as well, like `init(([Self])->Void)` — these are allowed too.